test/bench_sub_async.c

```diff
@@ -0,0 +1,500 @@
+                s = NATS_ERR;
+                break;
+            }
+            if ((int)(env->subs[i].count) != env->numPubMessages)
+            {
+                fprintf(stderr, "Error: count is %" PRId64 " for sub %d, expected %d\n", env->subs[i].count, i, env->numPubMessages);
+                s = NATS_ERR;
+                break;
+            }
+
+            int64_t dur = env->subs[i].closedTimestamp - start;
+            if (dur > w)
+                w = dur;
+            if ((dur < b) || (b == 0))
+                b = dur;
+            a += dur;
+        }
+    }
+
+    // cleanup
+    for (int i = 0; i < env->numSubs; i++)
+        natsSubscription_Destroy(env->subs[i].sub);
+    natsConnection_Destroy(nc);
+    natsOptions_Destroy(opts);
+    nats_CloseAndWait(0);
+
+    *best = (int)b;
+    *avg = (int)(a / env->numSubs);
+    *worst = (int)w;
+
+    _stopServer(pid);
+
+    return s;
+}
+
+static void _onMessage(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *closure)
+{
+    subState *ss = (subState *)closure;
+
+#ifndef _WIN32
+    if (ss->env->delayNano > 0)
+    {
+        struct timespec wait = {0, ss->env->delayNano};
+        nanosleep(&wait, NULL);
+    }
+#endif
+
+    char buf[32];
+    int len = natsMsg_GetDataLength(msg);
+    if (len > 31)
+        len = 31;
+
+    strncpy(buf, natsMsg_GetData(msg), len);
+    buf[len] = '\0';
+    int64_t val = atoi(buf);
+
+    ss->sum += val;
+    ss->xor ^= val;
+    ss->count++;
+
+    natsMsg_Destroy(msg);
+}
+
+static void _onComplete(void *closure)
+{
+    subState *ss = (subState *)closure;
+    ss->closedTimestamp = nats_Now();
+}
+
+static natsStatus _publish(natsConnection *nc, const char *subject, ENV *env)
+{
+    natsStatus s = NATS_OK;
+    char buf[16];
+
+    int flushAfter = env->progressiveFlush ? env->numPubMessages / (env->numSubs * 2) : // trigger
+                         env->numPubMessages + 1;                                       // do not trigger
+    for (int i = 0; i < env->numPubMessages; i++)
+    {
+        snprintf(buf, sizeof(buf), "%d", i);
+        IFOK(s, natsConnection_PublishString(nc, subject, buf));
+
+        if (((i != 0) && ((i % flushAfter) == 0)) || // progressive flush
+            (i == (env->numPubMessages - 1)))        // last message in batch
+        {
+            IFOK(s, natsConnection_Flush(nc));
+        }
+    }
+
+    return s;
+}
+
+static natsStatus _inject(natsConnection *nc, const char *subject, ENV *env)
+{
+    natsStatus s = NATS_OK;
+    natsMsg *m = NULL;
+    char buf[16];
+
+    for (int i = 0; i < env->numPubMessages; i++)
+    {
+        for (int n = 0; n < env->numSubs; n++)
+        {
+            snprintf(buf, sizeof(buf), "%d", i);
+
+            s = natsMsg_Create(&m, subject, NULL, buf, strlen(buf));
+            IFOK(s, _enqueueToSub(env->subs[n].sub, m));
+        }
+    }
+
+    return s;
+}
+
+static natsStatus _enqueueToSub(natsSubscription *sub, natsMsg *m)
```

Comment:
I see why we would have this (to test the library without the issue with server delivery), but this may need to be maintained in addition to the rest of the library code.

Reply:
In https://github.com/nats-io/nats.c/pull/772 there's already a `nats_enqueueMsgToSub`, so this will hopefully disappear soon. For now this was cut&pasted from `natsConn_processMsg`, the relevant path.